### PR TITLE
Support non-lm_head output projections in chunked SFT loss (GPTNeoX)

### PR DESCRIPTION
**Pull request by qgallouedec**
Allow `loss_type="chunked_nll"` to work with models whose output projection is not named `lm_head`.

The patcher now resolves the head via the public `PreTrainedModel.get_output_embeddings()` API instead of accessing `model.lm_head` directly. This unblocks GPTNeoX (`embed_out`) and any other standard `AutoModelForCausalLM` that happens to use a different attribute name.

Though, note that `lm_head` has been standard name for a long time.

## Why

On `main`, training a GPTNeoX model with the default SFT loss raises:

```python
from datasets import load_dataset
from trl import SFTTrainer, SFTConfig

trainer = SFTTrainer(
    model="trl-internal-testing/tiny-GPTNeoXForCausalLM",
    args=SFTConfig(loss_type="chunked_nll"),
    train_dataset=load_dataset("trl-internal-testing/zen", "standard_language_modeling", split="train"),
)
trainer.train()
```

```
AttributeError: 'GPTNeoXForCausalLM' object has no attribute 'lm_head'
```

`get_output_embeddings()` is the transformers contract for retrieving the LM head regardless of the underlying attribute name (`lm_head`, `embed_out`, …), so the patcher works uniformly across architectures.

## Tests

I'm not entirely convinced it's worth adding a new test for GPTNeoX specifically × chunked loss, it's more of an edge case, but happy to add one if you think it's necessary.

---

> [!NOTE]
> **Low Risk**
> Narrow change to head resolution for an optional loss path; aligns with the transformers API and should not affect models that already use `lm_head`.
> 
> **Overview**
> Fixes **`loss_type="chunked_nll"`** for causal LMs whose output projection is not exposed as **`lm_head`** (e.g. GPTNeoX’s **`embed_out`**).
> 
> The chunked-CE patcher and PEFT guard now resolve the head via **`PreTrainedModel.get_output_embeddings()`** instead of **`model.lm_head`**, and the patched forward uses that module’s **`.weight`** / **`.bias`** for chunked matmuls. Behavior for standard **`lm_head`** models and the existing “no PEFT on the output head” rule are unchanged; only the lookup path is generalized.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0b42395c999f3a04aeab599dd550127ceec7e846. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1259,18 +1260,17 @@ def __init__(
...
                         raise ValueError(
-                            "`loss_type='chunked_nll'` is not supported when `lm_head` is wrapped by a PEFT "
-                            "adapter (e.g. `target_modules='all-linear'` or explicitly including `'lm_head'`). "
-                            "Either remove `lm_head` from `target_modules`, or switch to `loss_type='nll'`. If "
-                            "this is a real use case for you, please open an issue at "
-                            "https://github.com/huggingface/trl/issues."
+                            "`loss_type='chunked_nll'` is not supported when `lm_head` is wrapped by a PEFT adapter "
+                            "(e.g. `target_modules='all-linear'` or explicitly including `'lm_head'`). Either remove "
+                            "`lm_head` from `target_modules`, or switch to `loss_type='nll'`. If this is a real use "
+                            "case for you, please open an issue at https://github.com/huggingface/trl/issues."
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:1273]: intended, I don't want to advertise this old feature, and promote a unified naming `lm_head`

**Review by kashif — APPROVED** [review]
